### PR TITLE
Optimize the HStore parser

### PR DESCRIPTION
**Pull request by casperisfine**
### Context

I saw some people reporting poor Hstore performance on Reddit, and after taking a quick look, it indeed seems that the parser is quite wasteful.

### Compatibility

A good part of the extra performance is at the cost of making the parser less tolerant. Maybe extra tests based on the postgres documentation were added by @joelhoffman in https://github.com/rails/rails/pull/4896, but from my understanding, the documentation states what postgres will accept, not what it will produce.

e.g. unquoted `NULL` is case insensitive, so postgress will accept `NuLL` instead, but will always return `NULL`.

I'm not experienced with postgresql though, so I'm might be totally off.

### Performance

Test string from taken from the the test suite: `"\"a\\\\b\"=>\"b\\\\ar\", \"1\\\"foo\"=>\"2\""`

Bench:

```
Warming up --------------------------------------
            original     6.896k i/100ms
             patched    15.787k i/100ms
Calculating -------------------------------------
            original     68.176k (± 1.2%) i/s -    344.800k in   5.058270s
             patched    157.786k (± 1.0%) i/s -    789.350k in   5.003144s

Comparison:
             patched:   157786.0 i/s
            original:    68176.1 i/s - 2.31x  (± 0.00) slower

```

Memory before: allocated 4376 bytes (55 objects)
Memory after: allocated: 1880 bytes (23 objects)

### CODE REVIEW

**Comment by casperisfine** [comment]
Thanks to @pixeltrix I pushed another version using `StringScanner` that's even slightly faster: https://gist.github.com/pixeltrix/a9aafda5d4d8bedf709ff54636551986

I'll merge on green.

**Comment by pixeltrix** [comment]
@casperisfine did you leave the `HstorePair` constant in for a reason?

**Comment by casperisfine** [comment]
No, that was a total oversight...

File: activerecord/lib/active_record/connection_adapters/postgresql/oid/hstore.rb [hunk 1]
```diff
@@ -1,26 +1,71 @@
...
+
           include ActiveModel::Type::Helpers::Mutable
 
           def type
             :hstore
           end
 
           def deserialize(value)
-            if value.is_a?(::String)
-              ::Hash[value.scan(HstorePair).map { |k, v|
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/postgresql/oid/hstore.rb:16]: doesn't look like `HstorePair` is referenced anywhere anymore. could it be removed?

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/hstore.rb:16]: I was in another commit on master.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql/oid/hstore.rb:16]: oops just saw the same comment above

**Pull request merged**